File: desktop/src/network.ts
export const USER_AGENT = "ISBN-to-Zotero/1.2 (local personal bibliographic resolver)";

export function applicationUserAgent(contact = ""): string {
  const cleaned = String(contact).trim().replace(/\s+/g, " ").slice(0, 200);
  if (!cleaned) return USER_AGENT;
  return `ISBN-to-Zotero/1.2 (local personal bibliographic resolver; contact: ${cleaned})`;
}

export class RequestPacer {
  readonly minimumIntervalMs: number;
  private queue: Promise<void> = Promise.resolve();
  private nextStart = 0;

  constructor(minimumIntervalMs: number) {
    this.minimumIntervalMs = Math.max(Number(minimumIntervalMs) || 0, 0);
  }

  async wait(): Promise<void> {
    let release = () => {};
    const previous = this.queue;
    this.queue = new Promise<void>(resolve => { release = resolve; });
    await previous;
    try {
      const delay = this.nextStart - Date.now();
      if (delay > 0) await Bun.sleep(delay);
      this.nextStart = Date.now() + this.minimumIntervalMs;
    } finally {
      release();
    }
  }
}

export class FetchError extends Error {
  url: string;
  status: number | null;
  temporary: boolean;

  constructor(url: string, message: string, status: number | null = null, temporary = false) {
    super(message);
    this.name = "FetchError";
    this.url = url;
    this.status = status;
    this.temporary = temporary;
  }
}

function sourceMessage(status: number): string {
  if (status === 404) return "No record at this source";
  if (status === 429) return "Source rate limit reached";
  return `HTTP ${status} from source`;
}

export async function fetchText(
  url: string,
  accept = "text/plain,*/*",
  timeoutMs = 18_000,
  userAgent = USER_AGENT,
): Promise<string> {
  let response: Response;
  try {
    response = await fetch(url, {
      headers: { "User-Agent": userAgent, Accept: accept },
      signal: AbortSignal.timeout(timeoutMs),
      redirect: "follow",
    });
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    throw new FetchError(url, `Source unavailable: ${detail}`, null, true);
  }
  if (!response.ok) {
    const temporary = [408, 425, 429, 500, 502, 503, 504].includes(response.status);
    throw new FetchError(url, sourceMessage(response.status), response.status, temporary);
  }
  return response.text();
}

export async function fetchJSON(url: string, timeoutMs = 18_000, userAgent = USER_AGENT): Promise<Record<string, unknown>> {
  const text = await fetchText(url, "application/json", timeoutMs, userAgent);
  try {
    const value = JSON.parse(text);
    if (!value || typeof value !== "object" || Array.isArray(value)) {
      throw new Error("not an object");
    }
    return value as Record<string, unknown>;
  } catch (error) {
    if (error instanceof FetchError) throw error;
    throw new FetchError(url, "Source returned malformed JSON");
  }
}

export async function mapLimit<T, R>(values: T[], limit: number, operation: (value: T, index: number) => Promise<R>): Promise<R[]> {
  if (!values.length) return [];
  const output = new Array<R>(values.length);
  let cursor = 0;
  const workers = Array.from({ length: Math.min(Math.max(limit, 1), values.length) }, async () => {
    while (cursor < values.length) {
      const index = cursor;
      cursor += 1;
      output[index] = await operation(values[index], index);
    }
  });
  await Promise.all(workers);
  return output;
}

File: python/isbn_zotero/network.py
from __future__ import annotations

import json
import socket
import threading
import time
import urllib.error
import urllib.request
from dataclasses import dataclass


USER_AGENT = "ISBN-to-Zotero/1.2 (local personal bibliographic resolver)"


def application_user_agent(contact: str = "") -> str:
    cleaned = " ".join(str(contact or "").split())[:200]
    if not cleaned:
        return USER_AGENT
    return f"ISBN-to-Zotero/1.2 (local personal bibliographic resolver; contact: {cleaned})"


class RequestPacer:
    def __init__(self, minimum_interval_seconds: float):
        self.minimum_interval_seconds = max(float(minimum_interval_seconds), 0.0)
        self.lock = threading.Lock()
        self.next_start = 0.0

    def wait(self) -> None:
        with self.lock:
            delay = self.next_start - time.monotonic()
            if delay > 0:
                time.sleep(delay)
            self.next_start = time.monotonic() + self.minimum_interval_seconds


@dataclass(slots=True)
class FetchError(Exception):
    url: str
    message: str
    status: int | None = None
    temporary: bool = False

    def __str__(self) -> str:
        return self.message


class HTTPClient:
    def __init__(self, timeout: float = 18.0):
        self.timeout = timeout

    def get_bytes(self, url: str, accept: str = "*/*", user_agent: str | None = None) -> bytes:
        request = urllib.request.Request(
            url,
            headers={"User-Agent": user_agent or USER_AGENT, "Accept": accept},
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return response.read()
        except urllib.error.HTTPError as error:
            temporary = error.code in {408, 425, 429, 500, 502, 503, 504}
            if error.code == 404:
                message = "No record at this source"
            elif error.code == 429:
                message = "Source rate limit reached"
            else:
                message = f"HTTP {error.code} from source"
            raise FetchError(url, message, error.code, temporary) from error
        except (urllib.error.URLError, TimeoutError, socket.timeout) as error:
            raise FetchError(url, f"Source unavailable: {error}", temporary=True) from error

    def get_text(self, url: str, accept: str = "text/plain,*/*", user_agent: str | None = None) -> str:
        return self.get_bytes(url, accept, user_agent=user_agent).decode("utf-8", "replace")

    def get_json(self, url: str, user_agent: str | None = None) -> dict:
        try:
            return json.loads(self.get_text(url, "application/json", user_agent=user_agent))
        except json.JSONDecodeError as error:
            raise FetchError(url, "Source returned malformed JSON") from error
